src/Components/Components/src/Routing/QueryParameterValueSupplier.cs

```diff
@@ -0,0 +1,186 @@
+                    builder.AddAttribute(0, destination.ComponentParameterName, blankValue);
+                }
+                return;
+            }
+
+            // Temporary workspace in which we accumulate the data while walking the querystring.
+            var valuesByDestination = ArrayPool<StringSegmentAccumulator>.Shared.Rent(_destinationsCount);
+
+            try
+            {
+                // Capture values by destination in a single pass through the querystring
+                var queryStringEnumerable = new QueryStringEnumerable(queryString);
+                foreach (var suppliedPair in queryStringEnumerable)
+                {
+                    if (_destinationsByQueryParameterName.TryGetValue(suppliedPair.EncodedName, out var destinations))
+                    {
+                        var decodedValue = suppliedPair.DecodeValue();
+
+                        foreach (ref var destination in CollectionsMarshal.AsSpan(destinations))
+                        {
+                            if (destination.IsArray)
+                            {
+                                valuesByDestination[destination.Index].Add(decodedValue);
+                            }
+                            else
+                            {
+                                valuesByDestination[destination.Index].SetSingle(decodedValue);
+                            }
+                        }
+                    }
+                }
+
+                // Finally, emit the parameter attributes by parsing all the string segments and building arrays
+                for (var destinationIndex = 0; destinationIndex < _destinations.Length; destinationIndex++)
+                {
+                    ref var destination = ref _destinations[destinationIndex];
+                    ref var values = ref valuesByDestination[destination.Index];
+
+                    var parsedValue = destination.IsArray
+                        ? destination.Parser.ParseMultiple(values, destination.ComponentParameterName)
+                        : values.Count == 0
+                            ? default
+                            : destination.Parser.Parse(values[0].Span, destination.ComponentParameterName);
+
+                    builder.AddAttribute(0, destination.ComponentParameterName, parsedValue);
+                }
+            }
+            finally
+            {
+                ArrayPool<StringSegmentAccumulator>.Shared.Return(valuesByDestination, true);
+            }
+        }
+
+        private static SortedDictionary<ReadOnlyMemory<char>, List<QueryParameterDestination>>? BuildQueryParameterMappings(Type componentType, out int destinationsCount)
+        {
+            var candidateProperties = MemberAssignment.GetPropertiesIncludingInherited(componentType, ComponentProperties.BindablePropertyFlags);
+            SortedDictionary<ReadOnlyMemory<char>, List<QueryParameterDestination>>? result = null;
```

Comment:
What's the benefit of sorting?

Reply:
The lookups will be equivalent to a binary search, i.e., a few string-compare calls, rather than having to compute a hash of an arbitrarily-long string (if it was `Dictionary<T, U>`).

Reply:
Since doing the perf investigations in 5.0 I'm very suspicious about any dictionaries keyed by string, as the hashing can be super expensive relative to the rest of the process.

Reply:
One of the things we do in MVC is to use List.BinarySearch / Insertion sort for this. https://github.com/dotnet/aspnetcore/blob/52eff90fbcfca39b7eb58baad597df6a99a542b0/src/Mvc/Mvc.TagHelpers/src/GlobbingUrlBuilder.cs#L171-L177. Is that an option here?

Reply:
I actually did implement it like that at one point but `SortedDictionary` is a more high-level, built-in way to do it. What's your concern with using it? Is it that we'll end up shipping more post-linked code? I didn't do a size comparison.

Reply:
I did a size comparison and found that the `SortedDictionary` implementation leads to about 2.8 KB more on the app size, after compression (compared with doing some more manual stuff with plain arrays).

So, I've removed `SortedDictionary` and switched to the slightly lower-level implementation to avert the size increase.